slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -279,7 +279,73 @@ ::grpc::Status SLAMServiceImpl::GetCurrentPointCloudMap(
 ::grpc::Status SLAMServiceImpl::GetPointCloudMapStream(
     ServerContext *context, const GetPointCloudMapStreamRequest *request,
     ServerWriter<GetPointCloudMapStreamResponse> *writer) {
-    return ::grpc::Status(::grpc::StatusCode::UNIMPLEMENTED, "");
+    std::string pointcloud_map;
+    int num_points;
+    // Write or grab the latest pointcloud map in form of a string
+    try {
+        std::shared_lock optimization_lock{optimization_shared_mutex,
+                                           std::defer_lock};
+        if (optimization_lock.try_lock()) {
+            // We are able to lock the optimization_shared_mutex, which means
+            // that the optimization is not ongoing and we can grab the newest
+            // map
+            num_points = GetLatestSampledPointCloudMapString(pointcloud_map);
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            latest_pointcloud_map = pointcloud_map;
+        } else {
+            // We couldn't lock the mutex which means the optimization process
+            // locked it and we need to use the backed up latest map
+            LOG(INFO)
+                << "Optimization is occuring, using cached pointcloud map";
+            std::lock_guard<std::mutex> lk(viam_response_mutex);
+            pointcloud_map = latest_pointcloud_map;
+        }
+    } catch (std::exception &e) {
+        LOG(ERROR) << "Stopping Cartographer: error encoding pointcloud: "
+                   << e.what();
+        std::terminate();
+    }
+
+    if (pointcloud_map.empty()) {
+        LOG(ERROR) << "map pointcloud does not have points yet";
+        return grpc::Status(grpc::StatusCode::UNAVAILABLE,
+                            "map pointcloud does not have points yet");
+    }
+
+    std::string pcd_chunk;
+    GetPointCloudMapStreamResponse response;
+
+    // Calculate chunk sizes to ensure no points data is split between chunks
+    int header_byte_size = viam::utils::pcdHeader(num_points, true).size();
+    int point_byte_size = 4 * sizeof(float);         // X, Y, Z, Color
+    int header_chunk_size =
+        std::floor((maximumGRPCByteChunkSize - header_byte_size) /
+                   point_byte_size) *
+        point_byte_size;
+    int chunk_size =
+        std::floor(maximumGRPCByteChunkSize / point_byte_size) *
+        point_byte_size;
+
+    // Send header chunk
+    pcd_chunk = pointcloud_map.substr(0, header_chunk_size);
+    response.set_point_cloud_pcd_chunk(pcd_chunk);
+    bool ok = writer->Write(response);
+    if (!ok)
+        return grpc::Status(
+            grpc::StatusCode::UNAVAILABLE,
+            "error while writing header to stream: stream closed");
+
+    // Send remaining point chunks
+    for (int start_index = header_chunk_size;
+         start_index < pointcloud_map.size(); start_index += chunk_size) {
```

Comment:
I think we won't end up sending the last points, if pointcloud_map.size() - header_chunk_size is not divisible by chunk_size.